examples/supplementary/arg_iter/src/lib.rs

```diff
@@ -0,0 +1,192 @@
+#![feature(rustc_private)]
+#![warn(unused_extern_crates)]
+
+extern crate rustc_hir;
+extern crate rustc_middle;
+extern crate rustc_span;
+
+use clippy_utils::diagnostics::span_lint_and_help;
+use rustc_hir::{Body, FnDecl, intravisit::FnKind};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_middle::ty::{self};
+use rustc_span::{
+    Span,
+    def_id::{DefId, LocalDefId},
+    symbol::Symbol,
+};
+
+dylint_linting::declare_late_lint! {
+    pub ARG_ITER,
+    Warn,
+    "functions taking `Iterator` trait bounds when `IntoIterator` would be more flexible"
+}
+
+impl<'tcx> LateLintPass<'tcx> for ArgIter {
+    fn check_fn(
+        &mut self,
+        cx: &LateContext<'tcx>,
+        _kind: FnKind<'tcx>,
+        decl: &'tcx FnDecl<'tcx>,
+        _body: &'tcx Body<'tcx>,
+        span: Span,
+        id: LocalDefId,
+    ) {
+        // Don't lint functions generated by macros (e.g. derive, attribute macros)
+        if span.from_expansion() {
+            return;
+        }
+
+        let Some(iterator_def_id) = cx.tcx.lang_items().iterator_trait() else {
+            return; // Can't find Iterator trait definition
+        };
+
+        // Use tcx.generics_of directly
+        let generics = cx.tcx.generics_of(id);
+
+        // No generics on this function
+        if generics.own_params.is_empty() {
+            return;
+        }
+
+        let predicates = cx.tcx.predicates_of(id);
+
+        // Find generic type parameters that are *only* bound by Iterator (+ implicit Sized)
+        let mut iterator_only_params: Vec<Symbol> = Vec::new(); // Store only the name (Symbol)
+        for param in &generics.own_params {
+            // Iterate over a slice
+            if let ty::GenericParamDefKind::Type { .. } = param.kind {
+                // param.def_id is already the DefId
+                if is_param_bound_only_by_iterator(cx, predicates, param.def_id, iterator_def_id) {
+                    iterator_only_params.push(param.name); // Push only the name
+                }
+            }
+        }
```

Comment:
I think I was not sufficiently clear. I think it should be possible to eliminate `iterator_only_params` and this loop. (See below.)